app/validation/ValidatedActions.scala

```diff
@@ -17,21 +16,32 @@ abstract class ValidatedActions() extends TokenSecurity with I18nSupport {
 
   def consignmentExists(consignmentId: UUID)(f: Request[AnyContent] => Result): Action[AnyContent] = secureAction.async {
     implicit request: Request[AnyContent] =>
-    val consignmentService = new ConsignmentService(graphqlConfiguration)
-    val consignmentExists = consignmentService.consignmentExists(consignmentId, request.token.bearerAccessToken)
-    consignmentExists.map {
-      case true => f(request)
-      case false => NotFound(views.html.notFoundError())
-    }
+      val consignmentService = new ConsignmentService(graphqlConfiguration)
+      val consignmentExists = consignmentService.consignmentExists(consignmentId, request.token.bearerAccessToken)
+      consignmentExists.map {
+        case true => f(request)
+        case false => NotFound(views.html.notFoundError())
+      }
   }
 
-  def transferAgreementExistsAction(consignmentId: UUID)(f: Request[AnyContent] => Result): Action[AnyContent]
-  = secureAction.async { implicit request: Request[AnyContent] =>
+  def uploadPermitted(consignmentId: UUID)(f: Request[AnyContent] => Result): Action[AnyContent] = secureAction.async {
+    implicit request: Request[AnyContent] =>
     val transferAgreementService = new TransferAgreementService(graphqlConfiguration)
+    val consignmentStatusService = new ConsignmentStatusService(graphqlConfiguration)
+
+    for {
+      transferAgreementExists <- transferAgreementService.transferAgreementExists(consignmentId, request.token.bearerAccessToken)
+      status <- consignmentStatusService.consignmentStatus(consignmentId, request.token.bearerAccessToken)
+    } yield {
+      val uploadInProgress: Boolean = status.flatMap(_.upload.map(_ == "InProgress")).getOrElse(false)
 
-    transferAgreementService.transferAgreementExists(consignmentId, request.token.bearerAccessToken) map {
-      case true => f(request)
-      case false => Redirect(routes.TransferAgreementController.transferAgreement(consignmentId))
+      if(!transferAgreementExists) {
+        Redirect(routes.TransferAgreementController.transferAgreement(consignmentId))
+      } else if(uploadInProgress) {
+        Redirect(routes.UploadController.uploadInProgress(consignmentId))
```

Comment:
This is interesting - I expected us to show the same page with a different message rather than doing it as a redirect.

What's the reason for doing it this way? I haven't thought about the trade-offs in any detail, so maybe this is the best option. My initial thought is that it's a bit odd to have a separate page because do you then have to redirect _back_ if someone happens to go there after the upload has finished?

Reply:
The reason I did it as a separate page is because the upload page is getting quite complicated already with different bits being shown or hidden depending on the state. This make accessibility more difficult as you have to make sure the page isn't dependant on css to show or hide things and we've already got bugs with things not showing when they should so I didn't want to make it worse. 
You're right about redirecting back when the upload is complete but we can add this logic into the controllers and it should be a bit clearer what's happening rather than digging through the javascript to see which bits are shown/hidden.

Reply:
Wait, wouldn't this logic be in the controller or HTML templates rather than the JavaScript and CSS?

This controller won't be any more complex if you render a different template on line 41 rather than do a redirect, and the user journeys will be easier to think about too because we don't need to add redirects on both controller actions.

Reply:
You're right about the html templates, ignore what I said about the css. I think I see what you mean. I thought you were talking about adding more if/else logic to the existing `upload.scala.html` which is what I was trying to avoid but do you mean rendering the `uploadInProgress.scala.html` template inside the validated actions? Like this 
```scala
if(!transferAgreementExists) {
        Redirect(routes.TransferAgreementController.transferAgreement(consignmentId))
      } else if(uploadInProgress) {
        Ok(views.html.uploadInProgress(consignmentId))
      } else {
        f(request)
      }
``` 
I've changed it to this as it makes sense I think

Reply:
Great, thanks!